text/0000-contract-metadata.md

```diff
@@ -0,0 +1,306 @@
+Suppose one wants to build a todo list app on blockchain, which is modeled as
+```typescript
+export class Todo {
+  id: string;
+  title: string;
+  completed: bool;
+}
+```
+
+and the contract is as follows
+```typescript
+// a contract that implements a todo list on blockchain
+import { context, storage, near, collections } from "./near";
+
+import { Todo } from "./model.near";
+
+// Map from string key ID to a Todo
+// collections.map is a persistent collection. Any changes to it will
+// be automatically saved in the storage.
+// The parameter to the constructor needs to be unique across a single contract.
+// It will be used as a prefix to all keys required to store data in the storage.
+let todos = collections.map<string, Todo>("todos");
+
+export function setTodo(id: string, todo: Todo): void {
+  near.log("setTodo " + id);
+  todos.set(id, todo);
+}
+
+@view
+export function getTodo(id: string): Todo {
+  return todos.get(id);
+}
+
+@view
+export function getAllTodos(): Array<Todo> {
+  // Map currently doesn't support getting all keys, so we use storage prefix.
+  let allKeys = storage.keys("todos::");
+  near.log("allKeys: " + allKeys.join(", "));
+
+  let loaded = new Array<Todo>(allKeys.length);
+  for (let i = 0; i < allKeys.length; i++) {
+    loaded[i] = Todo.decode(storage.getBytes(allKeys[i]));
+  }
+  return loaded;
+}
+
+@view
+export function description(): string {
+  return '{"name": "Todo list", "description": "A todo list on blockchain!"}'
+}
+```
+
+For this contract, `setTodo` is a change method while `getTodo` and `getAllTodos` are view methods.
+In this case, the metadata we want looks like 
+```json
+{
+  "methods": [
+               {
+                 "name": "setTodo",
+                 "parameters": [{"name": "id", "type": "string"}, {"name": "todo", "type": "Todo"}],
+                 "returnType": "void",
+                 "stateMutability": "change"
+               },
+               {
+                 "name": "getTodo",
+                 "parameters": [{"name": "id", "type": "string"}],
+                 "returnType": "Todo",
+                 "stateMutability": "view"
+               },
+               {
+                 "name": "getAllTodos",
+                 "parameters": [],
+                 "returnType": "Array<Todo>",
+                 "stateMutability": "view"
+               },
+               {
+                 "name": "description",
+                 "parameters": [],
+                 "returnType": "string",
+                 "stateMutability": "view"
+               }
+             ],
+   "classes": [
+                {
+                  "name": "Todo",
+                  "fields": [{"name": "id", "type": "string"}, {"name":  "title", "type": "string"}, {"name": "completed", "type": "bool"}]
+                }
+              ],
+   "contract": {"name":  "Todo list", "description": "A todo list on blockchain!"},
```

Comment:
contract name / description need to be localizable 

wonder if maybe it makes sense to keep it in separate metadata, so that user can call it like `description(userLocale)` and get this info for given locale.